### ISSUE TITLE
Tpetra: compile error in OpenMP-only build

### ISSUE THREAD

**Issue by brian-kelley** [trilinos/Trilinos#2495]
@trilinos/tpetra

## Current Behavior

On bowman with Intel 17.0.098 and OpenMPI 1.10.4, a release build of Tpetra with OpenMP as the only Node type fails. Enabling both Serial and OpenMP makes the error go away. This was on a fresh pull from develop (as of Mar. 29).

```
[ 77%] Building CXX object packages/tpetra/core/test/inout/CMakeFiles/TpetraCore_ReadTriples.dir/ReadTriples.cpp.o
/home/bmkelle/FreshTrilinos/Trilinos/packages/tpetra/core/inout/Tpetra_Details_PackTriples.hpp(364): error: static assertion failed with "countPackTriples: OrdinalType lacks an MpiTypeTraits specialization."
    static_assert (MpiTypeTraits<OrdinalType>::isSpecialized, "countPackTriples: "
    ^
          detected during:
            instantiation of "int Tpetra::Details::countPackTriples<ScalarType,OrdinalType>(int, const Teuchos::Comm<int> &, int &, std::ostream *) [with ScalarType=double, OrdinalType=<unnamed>::GO={long long}]" at line 689 of "/home/bmkelle/FreshTrilinos/Trilinos/packages/tpetra/core/inout/Tpetra_Details_ReadTriples.hpp"            instantiation of "int Tpetra::Details::Impl::readAndSendOneBatchOfTriples(std::istream &, std::size_t={unsigned long} &, std::size_t={unsigned long} &, Teuchos::ArrayRCP<int> &, Teuchos::ArrayRCP<char> &, std::vector<GO, std::allocator<GO>> &, std::vector<GO, std::allocator<GO>> &, std::vector<SC, std::allocator<SC>> &, std::size_t={unsigned long}, int, const Teuchos::Comm<int> &, bool, std::ostream *, bool) [with SC=double, GO=<unnamed>::GO={long long}]" at line 1019 of                      "/home/bmkelle/FreshTrilinos/Trilinos/packages/tpetra/core/inout/Tpetra_Details_ReadTriples.hpp"
            instantiation of "int Tpetra::Details::readAndDealOutTriples(std::istream &, std::size_t={unsigned long} &, std::size_t={unsigned long} &, std::function<int (GO, GO, const SC &)>, std::size_t={unsigned long}, const Teuchos::Comm<int> &, bool, std::ostream *, bool) [with SC=double, GO=<unnamed>::GO={long long}]" at line 210 of "/home/bmkelle/FreshTrilinos/Trilinos/packages/tpetra/core/test/inout/ReadTriples.cpp"

/home/bmkelle/FreshTrilinos/Trilinos/packages/tpetra/core/inout/Tpetra_Details_PackTriples.hpp(368): error #165: too few arguments in function call
    MPI_Datatype ordinalDt = MpiTypeTraits<OrdinalType>::getType ();
                                                                  ^
          detected during:
            instantiation of "int Tpetra::Details::countPackTriples<ScalarType,OrdinalType>(int, const Teuchos::Comm<int> &, int &, std::ostream *) [with ScalarType=double, OrdinalType=<unnamed>::GO={long long}]" at line 689 of "/home/bmkelle/FreshTrilinos/Trilinos/packages/tpetra/core/inout/Tpetra_Details_ReadTriples.hpp"
            instantiation of "int Tpetra::Details::Impl::readAndSendOneBatchOfTriples(std::istream &, std::size_t={unsigned long} &, std::size_t={unsigned long} &, Teuchos::ArrayRCP<int> &, Teuchos::ArrayRCP<char> &, std::vector<GO, std::allocator<GO>> &, std::vector<GO, std::allocator<GO>> &, std::vector<SC, std::allocator<SC>> &, std::size_t={unsigned long}, int, const Teuchos::Comm<int> &, bool, std::ostream *, bool) [with SC=double, GO=<unnamed>::GO={long long}]" at line 1019 of                      "/home/bmkelle/FreshTrilinos/Trilinos/packages/tpetra/core/inout/Tpetra_Details_ReadTriples.hpp"
```

## Steps to Reproduce
```
# in ../Trilinos/
git checkout f5753969053681ce7ec8d79f8aa0d159de59f35e
# in ../build/, put the module/configure script linked below
source config.sh
make -j8
```

## Your Environment

- Trilinos: f5753969053681ce7ec8d79f8aa0d159de59f35e
- bowman KNL testbed with icc 17.0.098
- [Module loading and Trilinos configure script](https://gist.github.com/brian-kelley/a27e8e6f8f71efa7dbb6cfe1499c8d18)
- Enabling Serial in the above configure script makes the build work

## Additional Information

[Full make output](https://gist.github.com/brian-kelley/40fb0368d081419e0e28f4d561aee43e)

**Comment by mhoemmen** [comment]
@brian-kelley Your build script sets `Teuchos_ENABLE_LONG_LONG_INT:BOOL=OFF`.  That option should always be `ON`, as Teuchos has required C++11 for several months at least.  C++11 comes with `long long`.

Also, there is no more "TpetraKernels" subpackage (though that option should just emit a harmless warning).

**Comment by mhoemmen** [comment]
I just tested debug gcc 4.9.3 with the following relevant build options:
```
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON
-D BUILD_SHARED_LIBS:BOOL=ON
-D Trilinos_ENABLE_OpenMP:BOOL=ON
-D Kokkos_ENABLE_OpenMP:BOOL=ON
-D Tpetra_INST_OPENMP:BOOL=ON
-D Trilinos_SHOW_DEPRECATED_WARNINGS:BOOL=ON
-D Trilinos_ENABLE_Fortran:BOOL=ON
-D CMAKE_CXX_FLAGS:STRING="-Wall --pedantic"
-D Teuchos_ENABLE_COMPLEX:BOOL=ON
-D Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON
-D TPL_ENABLE_MKL:BOOL=OFF
-D TPL_ENABLE_Boost:BOOL=ON
  -D Boost_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.59.0/gcc/4.9.3/base/include
  -D Boost_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.59.0/gcc/4.9.3/base/lib
-D TPL_ENABLE_BoostLib:BOOL=ON
  -D BoostLib_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.59.0/gcc/4.9.3/base/include
  -D BoostLib_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.59.0/gcc/4.9.3/base/lib
-D TPL_ENABLE_yaml-cpp:BOOL=OFF
-D TPL_ENABLE_MPI:BOOL=ON
-D CMAKE_BUILD_TYPE:STRING=DEBUG
-D Kokkos_ENABLE_Pthread:BOOL=OFF
-D Kokkos_ENABLE_Serial:BOOL=OFF
-D Tpetra_INST_SERIAL:BOOL=OFF
-D Kokkos_ENABLE_DEBUG:BOOL=ON
-D Teuchos_ENABLE_DEBUG:BOOL=ON
```
and Tpetra built just fine and passed all its tests.

**Comment by mhoemmen** [comment]
@brian-kelley The build error you noticed refers to `GO = long long`; you can't disable Teuchos' `long long` support and use `GO = long long` in Tpetra.  In fact, Tpetra should have emitted the following warning at configure time:
```
Do NOT set Teuchos_ENABLE_LONG_LONG_INT=OFF !  The default is ON.  We will VERY SOON require that this option be ON.  Don't set it explicitly!  Just use the default value.  See #1184: https://github.com/trilinos/Trilinos/issues/1184 and see also #1183: https://github.com/trilinos/Trilinos/issues/1183"
```

**Comment by mhoemmen** [comment]
@brian-kelley Don't worry!  I've opened #2496 to prevent issues like this in the future.  If you wish to take that issue please go ahead :-) .

**Comment by mhoemmen** [comment]
PR #2499, just merged into develop, fixes #2498.
